packages/dynamodb-emulator/index.js

```diff
@@ -0,0 +1,251 @@
+  if (options.inMemory && !options.dbPath) {
+    args.push('-inMemory');
+  }
+  if (options.sharedDb) {
+    args.push('-sharedDb');
+  }
+  return args;
+}
+
+async function launch(givenOptions = {}, retry = 0, startTime = Date.now()) {
+  log.info('launching', { retry, givenOptions });
+  // launch will retry but ensure it will not retry indefinitely.
+  if (retry >= maxRetries) {
+    throw new Error('max retries hit for starting dynamodb emulator');
+  }
+
+  if (givenOptions.inMemory && givenOptions.dbPath) {
+    throw new Error('inMemory and dbPath are mutually exclusive options');
+  }
+  let { port } = givenOptions;
+  if (!port) {
+    port = await portfinder.getPortPromise({ port: basePort });
+    log.info('found open port', { port });
+  }
+  const opts = { ...defaultOptions, ...givenOptions, port };
+
+  if (opts.dbPath) {
+    if (fs.existsSync(opts.dbPath)) {
+      const stat = fs.statSync(opts.dbPath);
+      if (!stat.isDirectory()) {
+        throw new Error(`dbPath must be a directory (${opts.dbPath})`);
+      }
+    } else {
+      log.info('Creating directory', { dbPath: opts.dbPath });
+      fs.mkdirSync(opts.dbPath);
+    }
+  }
+
+  const java = await which('java');
+  const args = buildArgs(opts);
+  const proc = spawn(java, args, {
+    cwd: emulatorPath,
+  });
+
+  function startingTimeout() {
+    log.error('Failed to start within timeout');
+    // ensure process is halted.
+    proc.kill();
+    const err = new Error('start has timed out!');
+    err.code = 'timeout';
+    throw err;
+  }
+
+  async function checkProcess() {
+    const portOnPID = await portPid(port);
+    // we invoke this after verifying the port we are attempting to
+    // bind has bound and we can connect. The logic here is if the
+    // port is bound but we have exited we know for sure that we need
+    // to retry starting the emulator.
+    if (
+      proc.exitCode != null ||
+      // verify that _we_ bound the port rather than another process.
+      portOnPID.all.length === 0 ||
+      portOnPID.all.indexOf(proc.pid) === -1
+    ) {
+      log.error('Port bound but by another process ... time to retry');
+      // port is open but it's not our process...
+      const err = new Error('port taken');
+      err.code = 'port_taken';
+      throw err;
+    }
+  }
+
+  // define this now so we can use it later to remove a listener.
+  let prematureExit;
+  // This is a fairly complex set of logic to retry starting
+  // the emulator if it fails to start. We need this logic due
+  // to possible race conditions between when we find an open
+  // port and bind to it. This situation is particularly common
+  // in jest where each test file is running in it's own process
+  // each competing for the open port.
+  try {
+    const waiter = wait(opts.startTimeout);
+    await Promise.race([
+      new Promise(accept => {
+        function readStdoutBuffer(buffer) {
+          if (buffer.toString().indexOf(opts.port)) {
```

Comment:
Do you want to:

`if (buffer.toString().indexOf(opts.port) !== -1` ?

Reply:
! now I am confused why this even works